File: intellij-plugin-structure/structure-classes/src/main/java/com/jetbrains/plugin/structure/classes/resolvers/PluginResolver.java
package com.jetbrains.plugin.structure.classes.resolvers;

import com.google.common.base.Throwables;
import com.jetbrains.plugin.structure.base.utils.FileUtil;
import com.jetbrains.plugin.structure.classes.locator.ClassesDirectoryLocator;
import com.jetbrains.plugin.structure.classes.locator.CompileServerExtensionLocator;
import com.jetbrains.plugin.structure.classes.locator.IdePluginClassesLocator;
import com.jetbrains.plugin.structure.classes.locator.LibDirectoryLocator;
import com.jetbrains.plugin.structure.intellij.extractor.*;
import com.jetbrains.plugin.structure.intellij.plugin.IdePlugin;
import com.jetbrains.plugin.structure.intellij.utils.StringUtil;
import org.apache.commons.io.IOUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.objectweb.asm.tree.ClassNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * @author Sergey Patrikeev
 */
public class PluginResolver extends Resolver {

  private static final Logger LOG = LoggerFactory.getLogger(PluginResolver.class);
  private static final List<IdePluginClassesLocator> CLASSES_LOCATORS = Arrays.asList(new ClassesDirectoryLocator(), new LibDirectoryLocator(), new CompileServerExtensionLocator());
  private final IdePlugin myPlugin;
  private final ExtractedPluginFile myExtractedPluginFile;
  private final Resolver myResolver;
  private boolean isClosed;

  private PluginResolver(IdePlugin plugin, @NotNull ExtractedPluginFile extractedPluginFile) {
    myPlugin = plugin;
    myExtractedPluginFile = extractedPluginFile;
    try {
      myResolver = loadClasses(myExtractedPluginFile.getActualPluginFile());
    } catch (Throwable e) {
      IOUtils.closeQuietly(extractedPluginFile);
      throw Throwables.propagate(e);
    }
  }

  @NotNull
  public static Resolver createPluginResolver(@NotNull IdePlugin plugin, @NotNull File extractDirectory) throws IOException {
    File pluginFile = plugin.getOriginalFile();
    if (pluginFile == null) {
      return Resolver.getEmptyResolver();
    } else if (!pluginFile.exists()) {
      throw new IllegalArgumentException("Plugin file doesn't exist " + pluginFile);
    }
    if (pluginFile.isDirectory() || FileUtil.INSTANCE.isJarOrZip(pluginFile)) {
      if (FileUtil.INSTANCE.isZip(pluginFile)) {
        ExtractorResult extractorResult = PluginExtractor.INSTANCE.extractPlugin(pluginFile, extractDirectory);
        if (extractorResult instanceof ExtractorSuccess) {
          ExtractedPluginFile extractedPluginFile = ((ExtractorSuccess) extractorResult).getExtractedPlugin();
          return new PluginResolver(plugin, extractedPluginFile);
        } else {
          throw new IOException(((ExtractorFail) extractorResult).getPluginProblem().getMessage());
        }
      }
      return new PluginResolver(plugin, new ExtractedPluginFile(pluginFile, null));
    }
    throw new IllegalArgumentException("Incorrect plugin file type " + pluginFile + ": expected a directory, a .zip or a .jar archive");
  }

  @Override
  public synchronized void close() throws IOException {
    if (isClosed) {
      return;
    }
    isClosed = true;

    try {
      myResolver.close();
    } finally {
      myExtractedPluginFile.close();
    }
  }

  @Nullable
  @Override
  public ClassNode findClass(@NotNull String className) throws IOException {
    return myResolver.findClass(className);
  }

  @Nullable
  @Override
  public Resolver getClassLocation(@NotNull String className) {
    return myResolver.getClassLocation(className);
  }

  @NotNull
  @Override
  public Iterator<String> getAllClasses() {
    return myResolver.getAllClasses();
  }

  @Override
  public boolean isEmpty() {
    return myResolver.isEmpty();
  }

  @Override
  public boolean containsClass(@NotNull String className) {
    return myResolver.containsClass(className);
  }

  @NotNull
  @Override
  public List<File> getClassPath() {
    return myResolver.getClassPath();
  }

  @NotNull
  private Resolver loadClasses(@NotNull File file) throws IOException {
    if (file.isDirectory()) {
      return loadClassesFromDir(file);
    } else if (file.exists() && StringUtil.endsWithIgnoreCase(file.getName(), ".jar")) {
      return createJarResolver(file);
    }
    throw new IllegalArgumentException("Invalid plugin file extension: " + file + ". It must be a directory or a jar file");
  }

  private Resolver loadClassesFromDir(@NotNull File pluginDirectory) {
    List<Resolver> resolvers = new ArrayList<Resolver>();
    try {
      for (IdePluginClassesLocator classesLocator : CLASSES_LOCATORS) {
        Resolver classes = classesLocator.findClasses(myPlugin, pluginDirectory);
        resolvers.add(classes);
      }
    } catch (Throwable e) {
      closeResolvers(resolvers);
      Throwables.propagate(e);
    }
    return createUnionResolver("Plugin resolver", resolvers);
  }

  private void closeResolvers(List<Resolver> resolvers) {
    for (Resolver resolver : resolvers) {
      try {
        resolver.close();
      } catch (Exception ce) {
        LOG.error("Unable to close resolver " + resolver, ce);
      }
    }
  }
}

File: intellij-plugin-structure/structure-classes/src/main/java/com/jetbrains/plugin/structure/classes/resolvers/PluginResolver.kt
package com.jetbrains.plugin.structure.classes.resolvers

import com.jetbrains.plugin.structure.base.utils.FileUtil
import com.jetbrains.plugin.structure.classes.locator.ClassesDirectoryLocator
import com.jetbrains.plugin.structure.classes.locator.CompileServerExtensionLocator
import com.jetbrains.plugin.structure.classes.locator.LibDirectoryLocator
import com.jetbrains.plugin.structure.intellij.extractor.ExtractedPluginFile
import com.jetbrains.plugin.structure.intellij.extractor.ExtractorFail
import com.jetbrains.plugin.structure.intellij.extractor.ExtractorSuccess
import com.jetbrains.plugin.structure.intellij.extractor.PluginExtractor
import com.jetbrains.plugin.structure.intellij.plugin.IdePlugin
import com.jetbrains.plugin.structure.intellij.utils.StringUtil
import org.apache.commons.io.IOUtils
import org.objectweb.asm.tree.ClassNode
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.util.*

/**
 * @author Sergey Patrikeev
 */
class PluginResolver private constructor(private val plugin: IdePlugin,
                                         private val extractedPluginFile: ExtractedPluginFile) : Resolver() {

  private val resolver: Resolver
  private var isClosed: Boolean = false

  init {
    try {
      resolver = loadClasses(extractedPluginFile.actualPluginFile)
    } catch (e: Throwable) {
      IOUtils.closeQuietly(extractedPluginFile)
      throw e
    }
  }

  @Synchronized
  override fun close() {
    if (isClosed) {
      return
    }
    isClosed = true

    extractedPluginFile.use {
      resolver.close()
    }
  }

  override fun findClass(className: String): ClassNode? = resolver.findClass(className)

  override fun getClassLocation(className: String): Resolver? = resolver.getClassLocation(className)

  override fun getAllClasses(): Iterator<String> = resolver.allClasses

  override fun isEmpty(): Boolean = resolver.isEmpty

  override fun containsClass(className: String): Boolean = resolver.containsClass(className)

  override fun getClassPath(): List<File> = resolver.classPath

  override fun getFinalResolvers(): List<Resolver> = resolver.finalResolvers

  private fun loadClasses(file: File): Resolver {
    if (file.isDirectory) {
      return loadClassesFromDir(file)
    } else if (file.exists() && StringUtil.endsWithIgnoreCase(file.name, ".jar")) {
      return Resolver.createJarResolver(file)
    }
    throw IllegalArgumentException("Invalid plugin file extension: $file. It must be a directory or a jar file")
  }

  private fun loadClassesFromDir(pluginDirectory: File): Resolver {
    val resolvers = arrayListOf<Resolver>()
    try {
      for (classesLocator in CLASSES_LOCATORS) {
        resolvers.addAll(classesLocator.findClasses(plugin, pluginDirectory))
      }
    } catch (e: Throwable) {
      closeResolvers(resolvers)
      throw e
    }
    val distinctResolvers = resolvers.flatMap { it.finalResolvers }.distinctBy { it.classPath }
    return Resolver.createUnionResolver("Plugin resolver", distinctResolvers)
  }

  private fun closeResolvers(resolvers: List<Resolver>) {
    for (resolver in resolvers) {
      try {
        resolver.close()
      } catch (ce: Exception) {
        LOG.error("Unable to close resolver " + resolver, ce)
      }
    }
  }

  companion object {
    private val LOG = LoggerFactory.getLogger(PluginResolver::class.java)

    private val CLASSES_LOCATORS = Arrays.asList(ClassesDirectoryLocator(), LibDirectoryLocator(), CompileServerExtensionLocator())

    @Throws(IOException::class)
    fun createPluginResolver(plugin: IdePlugin, extractDirectory: File): Resolver {
      val pluginFile = plugin.originalFile
      if (pluginFile == null) {
        return Resolver.getEmptyResolver()
      } else if (!pluginFile.exists()) {
        throw IllegalArgumentException("Plugin file doesn't exist " + pluginFile)
      }
      if (pluginFile.isDirectory || FileUtil.isJarOrZip(pluginFile)) {
        if (FileUtil.isZip(pluginFile)) {
          val extractorResult = PluginExtractor.extractPlugin(pluginFile, extractDirectory)
          if (extractorResult is ExtractorSuccess) {
            val extractedPluginFile = extractorResult.extractedPlugin
            return PluginResolver(plugin, extractedPluginFile)
          } else {
            throw IOException((extractorResult as ExtractorFail).pluginProblem.message)
          }
        }
        return PluginResolver(plugin, ExtractedPluginFile(pluginFile, null))
      }
      throw IllegalArgumentException("Incorrect plugin file type $pluginFile: expected a directory, a .zip or a .jar archive")
    }
  }
}
